### ISSUE TITLE
Visual Studio 17.2.6 Breaks CommunityToolkit.Mvvm

### ISSUE THREAD

**Issue by jamiehankins** [dotnet/roslyn#62582]
**Version Used**:  17.2.6

**Steps to Reproduce**:

1. Create a WPF exe project targeting `net6.0-windows`
2. Reference CommunityToolkit.Mvvm like so:
`<PackageReference Include="CommunityToolkit.Mvvm" Version="7.1.2" />`
4. Create a simple class inheriting from `ObservableObject`, like so:
```
using CommunityToolkit.Mvvm.ComponentModel;

namespace ObservablePropertyTest
{
    public partial class TestVM : ObservableObject
    {
        [ObservableProperty]
        private string _testString = string.Empty;
    }
}
```

**Expected Behavior**:
In 17.2.5, it built and ran fine.

**Actual Behavior**:
```
Rebuild started...
1>------ Rebuild All started: Project: ObservablePropertyTest, Configuration: Debug Any CPU ------
Restored C:\git\ObservablePropertyTest\ObservablePropertyTest.csproj (in 2 ms).
1>C:\git\ObservablePropertyTest\CommunityToolkit.Mvvm.SourceGenerators\CommunityToolkit.Mvvm.SourceGenerators.ObservablePropertyGenerator\__KnownINotifyPropertyChangedOrChangingArgs.cs(12,27,12,70): error CS0101: The namespace 'CommunityToolkit.Mvvm.ComponentModel.__Internals' already contains a definition for '__KnownINotifyPropertyChangedOrChangingArgs'
1>C:\git\ObservablePropertyTest\CommunityToolkit.Mvvm.SourceGenerators\CommunityToolkit.Mvvm.SourceGenerators.ObservablePropertyGenerator\__KnownINotifyPropertyChangedOrChangingArgs.cs(7,6,7,51): error CS0579: Duplicate 'global::System.CodeDom.Compiler.GeneratedCode' attribute
1>C:\git\ObservablePropertyTest\CommunityToolkit.Mvvm.SourceGenerators\CommunityToolkit.Mvvm.SourceGenerators.ObservablePropertyGenerator\__KnownINotifyPropertyChangedOrChangingArgs.cs(8,6,8,52): error CS0579: Duplicate 'global::System.Diagnostics.DebuggerNonUserCode' attribute
1>C:\git\ObservablePropertyTest\CommunityToolkit.Mvvm.SourceGenerators\CommunityToolkit.Mvvm.SourceGenerators.ObservablePropertyGenerator\__KnownINotifyPropertyChangedOrChangingArgs.cs(9,6,9,69): error CS0579: Duplicate 'global::System.Diagnostics.CodeAnalysis.ExcludeFromCodeCoverage' attribute
1>C:\git\ObservablePropertyTest\CommunityToolkit.Mvvm.SourceGenerators\CommunityToolkit.Mvvm.SourceGenerators.ObservablePropertyGenerator\__KnownINotifyPropertyChangedOrChangingArgs.cs(10,6,10,51): error CS0579: Duplicate 'global::System.ComponentModel.EditorBrowsable' attribute
1>C:\git\ObservablePropertyTest\CommunityToolkit.Mvvm.SourceGenerators\CommunityToolkit.Mvvm.SourceGenerators.ObservablePropertyGenerator\__KnownINotifyPropertyChangedOrChangingArgs.cs(11,6,11,29): error CS0579: Duplicate 'global::System.Obsolete' attribute
1>C:\git\ObservablePropertyTest\CommunityToolkit.Mvvm.SourceGenerators\CommunityToolkit.Mvvm.SourceGenerators.ObservablePropertyGenerator\ObservablePropertyTest.TestVM.cs(12,23,12,33): error CS0102: The type 'TestVM' already contains a definition for 'TestString'
1>Done building project "ObservablePropertyTest_yynlzhol_wpftmp.csproj" -- FAILED.
========== Rebuild All: 0 succeeded, 1 failed, 0 skipped ==========
```

It appears that somehow the CommunityToolkit.Mvvm source generators are being run twice.

If you create a WPF class library, things work fine. They're only broken in a WPF app.

Here is my test project:
[ObservablePropertyTest.zip](https://github.com/dotnet/roslyn/files/9097397/ObservablePropertyTest.zip)

**Comment by jamiehankins** [comment]
@RikkiGibson 
> If you build your project from the developer command prompt and pass `-bl` it should produce a binary log which you can use to confirm that:
> 
> 1. multiple source generators are being passed
> 2. where those generators originate from.

Wow, first I've seen of this tool. There's an insane amount of info, but I can't seem to find that needle in this haystack.

Here's the output from building my tiny test project in case you have a better idea of where in there to look.
[msbuild.zip](https://github.com/dotnet/roslyn/files/9098382/msbuild.zip)

**Comment by RikkiGibson** [comment]
It's in the Csc task which contains the build errors, under CommandLineArguments

![image](https://user-images.githubusercontent.com/5833655/178636972-a1d14f75-225d-46db-8d42-2a4d55949443.png)

In this case the paths appear to be:

- `C:\Users\jamie.hankins\.nuget\packages\communitytoolkit.mvvm\7.1.2\analyzers\dotnet\cs\CommunityToolkit.Mvvm.SourceGenerators.dll`
- `C:\Users\jamie.hankins\.nuget\packages\communitytoolkit.mvvm\7.1.2\analyzers\dotnet\cs\CommunityToolkit.Mvvm.SourceGenerators.dll`

In other words, the two paths are exactly the same. It's odd that msbuild would do that, It feels like the compiler should be modified to give a diagnostic and/or dedupe the duplicate generators. I'll have to look more closely later to confirm that and determine a root cause.

**Comment by DingpingZhang** [comment]
This is a generic problem that affects all source generator projects.
It seems to occur only when referenced by the nuget package, If referenced in the following (manual reference), there will be no problem, so the test cases within the source generator project are normal.

```xml
    <ProjectReference Include="..\Xxx.Generator\Xxx.Generator.csproj" OutputItemType="Analyzer" ReferenceOutputAssembly="false" />
```

**Comment by Tum4ik** [comment]
I also have faced the same problem with Refit.

**Comment by tan-force** [comment]
I uninstalled VS 17.2.6 and installed 17.2.5 that used to work. It doesn't anymore :-(  Still the same error.

**Comment by glennawatson** [comment]
@tan-force seems to be related to dotnet 6.0.7 so you may have to uninstall that version

**Comment by tan-force** [comment]
Yes! Uninstalled dotnet 6.0.7, and now it works. Thank you.

**Comment by Tum4ik** [comment]
You still can use the latest version of Visual Studio. Here is a workaround that works for me:
1. Download and install the previous version of .NET SDK (6.0.301) https://dotnet.microsoft.com/en-us/download/dotnet/6.0
2. Add `global.json` file to the root of your solution (where .sln file is located)
3. The content of the `global.json`:
```json
{
    "sdk": {
        "version": "6.0.301",
        "rollForward": "disable"
    }
}
```
4. Rebuild solution.

**Comment by aahz16384** [comment]
Source generators still working in actual VS2022 preview.

**Comment by LadislavLang** [comment]
I've the same problem in latest preview (17.3.0 Preview 3.0)

**Comment by candritzky** [comment]
We have the same problem with a WPF (class library) project that uses **Refit**:

```
...\InterfaceStubGeneratorV2\Refit.Generator.InterfaceStubGeneratorV2\PreserveAttribute.g.cs(8,18): error CS0101: The namespace 'MyWpfClassLib_tde0zyvz_wpftmpRefitInternalGenerated' already contains a definition for 'PreserveAttribute' [...\MyWpfClassLib\MyWpfClassLib_tde0zyvz_wpftmp.csproj]
...\InterfaceStubGeneratorV2\Refit.Generator.InterfaceStubGeneratorV2\PreserveAttribute.g.cs(5,6): error CS0579: Duplicate 'global::System.Diagnostics.CodeAnalysis.ExcludeFromCodeCoverage' attribute [...\MyWpfClassLib_tde0zyvz_wpftmp.csproj]
...\InterfaceStubGeneratorV2\Refit.Generator.InterfaceStubGeneratorV2\PreserveAttribute.g.cs(6,6): error CS0579: Duplicate 'global::System.ComponentModel.EditorBrowsable' attribute [...\MyWpfClassLib_tde0
zyvz_wpftmp.csproj]
...\InterfaceStubGeneratorV2\Refit.Generator.InterfaceStubGeneratorV2\PreserveAttribute.g.cs(7,6): error CS0579: Duplicate 'global::System.AttributeUsage' attribute [...\MyWpfClassLib_tde0zyvz_wpftmp.csproj]
...\InterfaceStubGeneratorV2\Refit.Generator.InterfaceStubGeneratorV2\Generated.g.cs(7,6): error CS0579: Duplicate 'global::System.Diagnostics.CodeAnalysis.ExcludeFromCodeCoverage' attribute [...\MyWpfClassLib_tde0zyvz_wpftmp.csproj]
...\InterfaceStubGeneratorV2\Refit.Generator.InterfaceStubGeneratorV2\Generated.g.cs(8,6): error CS0579: Duplicate 'global::System.Diagnostics.DebuggerNonUserCode' attribute [...\MyWpfClassLib_tde0zyvz_wpftmp.csproj]
...\InterfaceStubGeneratorV2\Refit.Generator.InterfaceStubGeneratorV2\Generated.g.cs(9,6): error CS0579: Duplicate 'global::MyWpfClassLib_tde0zyvz_wpftmpRefitInternalGenerated.PreserveAttribute' attribute [...\MyWpfClassLib_tde0zyvz_wpftmp.csproj]
...\InterfaceStubGeneratorV2\Refit.Generator.InterfaceStubGeneratorV2\Generated.g.cs(11,6): error CS0579: Duplicate 'global::System.ComponentModel.EditorBrowsable' attribute [...\MyWpfClassLib_tde0zyvz_wpftmp.csproj]
...\InterfaceStubGeneratorV2\Refit.Generator.InterfaceStubGeneratorV2\IVersionInfoRestClient.g.cs(10,6): error CS0579: Duplicate 'global::System.Diagnostics.CodeAnalysis.ExcludeFromCodeCoverage' attribute [...\MyWpfClassLib_tde0zyvz_wpftmp.csproj]
...\InterfaceStubGeneratorV2\Refit.Generator.InterfaceStubGeneratorV2\IVersionInfoRestClient.g.cs(11,6): error CS0579: Duplicate 'global::System.Diagnostics.DebuggerNonUserCode' attribute [...\MyWpfClassLib_tde0zyvz_wpftmp.csproj]
...\InterfaceStubGeneratorV2\Refit.Generator.InterfaceStubGeneratorV2\IVersionInfoRestClient.g.cs(12,6): error CS0579: Duplicate 'global::MyWpfClassLib_tde0zyvz_wpftmpRefitInternalGenerated.PreserveAttribute' attribute [...\MyWpfClassLib\MyWpfClassLib_tde0zyvz_wpftmp.csproj]
...\InterfaceStubGeneratorV2\Refit.Generator.InterfaceStubGeneratorV2\IVersionInfoRestClient.g.cs(14,6): error CS0579: Duplicate 'global::System.ComponentModel.EditorBrowsable' attribute [...\MyWpfClassLib_tde0zyvz_wpftmp.csproj]
...\InterfaceStubGeneratorV2\Refit.Generator.InterfaceStubGeneratorV2\IVersionInfoRestClient.g.cs(15,19): error CS8646: 'SettingsValidationService.IVersionInfoRestClient.GetConfigurationAsync(CancellationToken)' is explicitly implemented more than once. [...\MyWpfClassLib\MyWpfClassLib_tde0zyvz_wpftmp.csproj]
...\InterfaceStubGeneratorV2\Refit.Generator.InterfaceStubGeneratorV2\IVersionInfoRestClient.g.cs(15,19): error CS8646: 'SettingsValidationService.IVersionInfoRestClient.GetVersionInfoAsync(CancellationToken)' is explicitly implemented more than once. [...\MyWpfClassLib\MyWpfClassLib_tde0zyvz_wpftmp.csproj]
...\InterfaceStubGeneratorV2\Refit.Generator.InterfaceStubGeneratorV2\IVersionInfoRestClient.g.cs(20,51): error CS0102: The type 'Generated.MyClassLibIVersionInfoRestClient' already contains a definition for 'Client' [...\MyWpfClassLib\MyWpfClassLib_tde0zyvz_wpftmp.csproj]
...\InterfaceStubGeneratorV2\Refit.Generator.InterfaceStubGeneratorV2\IVersionInfoRestClient.g.cs(21,48): error CS0102: The type 'Generated.MyClassLibIVersionInfoRestClient' already contains a definition for 'requestBuilder' [...\MyWpfClassLib\MyWpfClassLib_tde0zyvz_wpftmp.csproj]
...\InterfaceStubGeneratorV2\Refit.Generator.InterfaceStubGeneratorV2\IVersionInfoRestClient.g.cs(24,16): error CS0111: Type 'Generated.MyClassLibIVersionInfoRestClient' already defines a member called 'MyClassLibIVersionInfoRestClient' with the same parameter types [...\MyWpfClassLib_tde0zyvz_wpftmp.csproj]
...\InterfaceStubGeneratorV2\Refit.Generator.InterfaceStubGeneratorV2\IVersionInfoRestClient.g.cs(33,52): error CS0111: Type 'Generated.MyClassLibIVersionInfoRestClient' already defines a member called 'GetConfigurationAsync' with the same parameter types [...\MyWpfClassLib_tde0zyvz_wpftmp.csproj]
...\InterfaceStubGeneratorV2\Refit.Generator.InterfaceStubGeneratorV2\IVersionInfoRestClient.g.cs(41,140): error CS0111: Type 'Generated.MyClassLibIVersionInfoRestClient' already defines a member called 'GetVersionInfoAsync' with the same parameter types [...\MyWpfClassLib_tde0zyvz_wpftmp.csproj]
...\InterfaceStubGeneratorV2\Refit.Generator.InterfaceStubGeneratorV2\IVersionInfoRestClient.g.cs(49,135): error CS0111: Type 'Generated.MyClassLibIVersionInfoRestClient' already defines a member called 'global::MyWpfClassLib.Settings.SettingsValidationService.IVersionInfoRestClient.GetConfigurationAsync' with the same parameter types [...\MyWpfClassLib_tde0zyvz_wpftmp.csproj]
...\InterfaceStubGeneratorV2\Refit.Generator.InterfaceStubGeneratorV2\IVersionInfoRestClient.g.cs(57,223): error CS0111: Type 'Generated.MyClassLibIVersionInfoRestClient' already defines a member called 'global::MyWpfClassLib.Settings.SettingsValidationService.IVersionInfoRestClient.GetVersionInfoAsync' with the same parameter types [...\MyWpfClassLib_tde0zyvz_wpftmp.csproj]
```

Something seems to be broken in the **GenerateTemporaryTargetAssembly** build task.

**Comment by CyrusNajmabadi** [comment]
I don't believe this is the  right repo for this issue.  @rainersigwald do you know where this may need to go?

**Comment by saivineeth100** [comment]
> You still can use the latest version of Visual Studio. Here is a workaround that works for me:
> 
> 1. Download and install the previous version of .NET SDK (6.0.301) https://dotnet.microsoft.com/en-us/download/dotnet/6.0
> 2. Add `global.json` file to the root of your solution (where .sln file is located)
> 3. The content of the `global.json`:
> 
> ```json
> {
>     "sdk": {
>         "version": "6.0.301",
>         "rollForward": "disable"
>     }
> }
> ```
> 
> 4. Rebuild solution.

Cannot Install SDK as it is showing error - Another version of this product is already installed

**Comment by ThomasGoulet73** [comment]
I apologize for the regression. It was caused by my changes in dotnet/wpf#6534 which were backported to .Net 6 in dotnet/wpf#6680.

I posted workaround for this issue here: https://github.com/dotnet/wpf/pull/6680#issuecomment-1183455941. This is an alternative to this workaround: https://github.com/dotnet/roslyn/issues/62582#issuecomment-1183251686.

I'm working on a fix that can hopefully be included in .Net 6.0.8.
